### PR TITLE
Support dynamic shapes in TritonTemplates

### PR DESCRIPTION
**Pull request by ipiszy**
Currently when dynamic=True, TritonTemplates won't be used, as the condition  `if list(call_args) != expected_args` defined in `TritonTemplate` cannot be satisfied. This PR tries to fix this issue by allowing passing symbolic variable names via `extra_args` and replacing all symbolic values in the generated TritonTemplate code as call_arg names. 

With this change, a locally compiled mm + epilogue node calls into the Triton kernel successfully.

This PR also introduces a new config "max_autotune_gemm_backends" to allow specifying candidate gemm backends for max autotune. Current choices: combinations of ATEN, TRITON. This makes tests easier, so that we can explicitly test Triton gemm kernels + epilogue fusions + dynamic shapes, without falling back to ATen ops.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Review by jansel — COMMENTED** [review]
Seems like a reasonable fix.
1) This is just changing how the kernel is called, not the generated code correct?
2) Can you add a test that exercises this code path?

File: torch/_inductor/select_algorithm.py [hunk 1]
```diff
@@ -197,6 +197,11 @@ def store_output(self, indices, val, mask):
                 V.graph.sizevars.simplify(s) for s in self.output_node.get_size()
             ]
             assert len(indices) == len(lengths)
+            replacements = {
+                x: self.args.size(x)
+                for x in index_symbols + lengths
+                if x.name.startswith("s") or x.name.startswith("ps")
```
Comment (aakhundov) [inline torch/_inductor/select_algorithm.py:203]: Curious if there is a more robust way of checking this. Or do the `s*` and `ps*` necessarily cover everything we want to have in the `replacements`? As written, I feel it's easy to miss some cases in the future. Maybe there is a helper function (or define one if not)?

**Comment by ipiszy** [comment]
> Seems like a reasonable fix.
> 
> 1. This is just changing how the kernel is called, not the generated code correct?

wrt generated code: The main template body handles "size" and "stride" for dynamic shapes correctly. However "store_output" part doesn't substitute symbolic symbols with variable names. This PR fixes this.

> 2. Can you add a test that exercises this code path?

Yeah will do.

File: torch/_inductor/select_algorithm.py (hunk 1 above)
Comment (ipiszy) [reply torch/_inductor/select_algorithm.py:203]: I copied it from https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/common.py#L928 haha. Let me extract it into a function.

File: torch/_inductor/select_algorithm.py [hunk 2]
```diff
@@ -197,6 +197,11 @@ def store_output(self, indices, val, mask):
                 V.graph.sizevars.simplify(s) for s in self.output_node.get_size()
             ]
             assert len(indices) == len(lengths)
+            replacements = {
+                x: self.args.size(x)
+                for x in index_symbols + lengths
+                if isinstance(x, sympy.Symbol) and is_size_var(x.name)
+            }
```
Comment (jansel) [inline torch/_inductor/select_algorithm.py:204]: I think self.rename_indexing() might do the same thing as this.

Comment (ipiszy) [reply torch/_inductor/select_algorithm.py:204]: Ah you're right. I thought rename_indexing() was only for single symbol replacement previously.

**Comment by ipiszy** [comment]
Updated the PR with unittests for max_autotune + dynamic shapes. Also introduced a new config "max_autotune_gemm_backends" to allow us explicitly testing Triton templates only.

File: torch/_inductor/kernel/mm.py [hunk 3]
```diff
@@ -143,26 +145,34 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
...
+                aten_addmm.bind(
+                    (inp, mat1, mat2),
+                    layout,
+                    ordered_kwargs_for_cpp_kernel,
+                    alpha=alpha,
+                    beta=beta,
+                )
+            ]
+            if use_aten_gemm_kernels()
+            else []
```
Comment (jansel) [inline torch/_inductor/kernel/mm.py:159]: Won't this result in passing an empty list to `autotune_select_algorithm` below?

File: torch/_inductor/kernel/mm_plus_mm.py [hunk 4]
```diff
@@ -158,7 +158,11 @@ def tuned_mm_plus_mm(mat1, mat2, mat3, mat4, *, layout=None):
 
     assert layout1 == layout2
     # options to tune from
-    choices = [aten_mm_plus_mm.bind((mat1, mat2, mat3, mat4), layout1)]
+    choices = (
+        [aten_mm_plus_mm.bind((mat1, mat2, mat3, mat4), layout1)]
+        if use_aten_gemm_kernels
+        else []
+    )
     if use_triton_template(layout1):
```
Comment (jansel) [inline torch/_inductor/kernel/mm_plus_mm.py:166]: If `use_triton_template()` is false (which happens for CPU kernels), then choices will be empty and we will hit an error.

Same in other files.

**Review by jansel — CHANGES_REQUESTED** [review]
We should have a better error message (explaining `config.max_autotune_gemm_backends`) for the case when no backends are valid.

File: torch/_inductor/kernel/bmm.py [hunk 5]
```diff
@@ -90,7 +90,7 @@ def tuned_bmm(mat1, mat2, *, layout=None):
     m, n, k, layout, mat1, mat2 = mm_args(mat1, mat2, layout=layout)
 
     # options to tune from
-    choices = [aten_bmm.bind((mat1, mat2), layout)]
+    choices = [aten_bmm.bind((mat1, mat2), layout)] if use_aten_gemm_kernels else []
```
Comment (aakhundov) [inline torch/_inductor/kernel/bmm.py:93]: Should be `use_aten_gemm_kernels()`? Also in few other place below.

**Review by ipiszy — COMMENTED** [review]
Thanks @jansel @aakhundov !

**Review by jansel — APPROVED** [review]

File: test/inductor/test_cpp_wrapper.py [hunk 6]
```diff
@@ -243,7 +243,13 @@ class BaseTest(NamedTuple):
         BaseTest("test_bmm1"),
         BaseTest("test_bmm2"),
         BaseTest("test_cat"),  # alias
-        BaseTest("test_convolution1"),
+
+        # TODO: Re-enable this test after fixing cuda wrapper for conv Triton templates with dynamic shapes.
```
Comment (ipiszy) [inline test/inductor/test_cpp_wrapper.py:247]: @desertfire FYI I disabled this test in cpp wrapper as dynamic shape doesn't work well for Conv Triton templates.

File: torch/_inductor/kernel/mm.py [hunk 7]
```diff
@@ -143,26 +145,34 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
...
-    choices = [
-        aten_addmm.bind(
-            (inp_expanded, mat1, mat2),
-            layout,
-            ordered_kwargs_for_cpp_kernel,
-            alpha=alpha,
-            beta=beta,
-        )
-    ]
+        if use_aten_gemm_kernels
```
Comment (jansel) [inline torch/_inductor/kernel/mm.py:173]:
```suggestion
        if use_aten_gemm_kernels()
```

Comment (ipiszy) [reply torch/_inductor/kernel/mm.py]: Sorry! Fixed it.

File: test/inductor/test_cpp_wrapper.py (hunk 6 above)
Comment (desertfire) [reply test/inductor/test_cpp_wrapper.py]: Ok to me.

**Pull request closed without merging**